### PR TITLE
add display commands

### PR DESCRIPTION
**Pull request by escoand**
Add commands to switch on and off HDMI displays (#1178).
Missing is the UI integration and the scheduling.

### CODE REVIEW

File: server.py [hunk 1]
```diff
@@ -190,6 +190,17 @@ def cleanup_usb_assets(media_dir='/media'):
                     if location.group() not in mountpoints:
                         assets_helper.delete(conn, asset['asset_id'])
 
+@celery.task
+def display_on():
+    cmd = sh.Command("/usr/bin/cec-client")
+    cmd(sh.echo('as 0'), '-d', '1', '-s')
```
Comment (vpetersson) [inline server.py]: Can you document what these arguments do?

**Comment by ealmonte32** [comment]
@escoand 
did the standby command work for you on the TV? i tested this as you can see https://github.com/Screenly/screenly-ose/issues/1178#issuecomment-553561107 and that command did not turn off the TV..

**Comment by escoand** [comment]
@ealmonte32 yes, on multiple devices

**Comment by ealmonte32** [comment]
> @ealmonte32 yes, on multiple devices

hmm.. during my tests:
```
pi@screenly400w:~ $ echo standby 0 | cec-client -s -d 1
opening a connection to the CEC adapter...
```
the TV remained on..

**Comment by escoand** [comment]
@ealmonte32 than maybe your device is not fully CEC compatible

**Comment by escoand** [comment]
@ealmonte32 have you tried different HDMI cables? I know there are some which are not able - the needed pin 13 is not always there.

**Comment by escoand** [comment]
@vpetersson I've added two buttons to assets page to turn the display on and off. I've currently no idea how to use a toggle button which gets updated according to the real display status.

For the settings page we need at least one checkbox to de-/activate and two time pickers. Do you prefer a specific position. Then I would do a proposal.

**Comment by ealmonte32** [comment]
> @ealmonte32 have you tried different HDMI cables? I know there are some which are not able - the needed pin 13 is not always there.

then the other commands wouldn't have received a response either so I dont think this is the case here.. it is a Hitachi Roku 4K TV, fairly new, hard to believe it doesnt support CEC.. plus, if it didn't, this wouldnt get this response:

```
pi@screenly400w:~ $ echo scan | cec-client -s -d 1
opening a connection to the CEC adapter...
requesting CEC bus information ...
CEC bus information
===================
device #0: TV
address:       0.0.0.0
active source: no
vendor:        Unknown
osd string:    TV
CEC version:   1.4
power status:  on
language:      eng
```

File: docker/Dockerfile.celery.template [hunk 2]
```diff
@@ -3,6 +3,7 @@ FROM balenalib/%%BALENA_MACHINE_NAME%%-debian:jessie
 RUN apt-get update && \
     apt-get -y install \
         build-essential \
+        cec-utils \
```
Comment (rusko124) [inline docker/Dockerfile.celery.template]: Could you add the dependency to `ansible/roles/system/tasks/main.yml`?
`docker/Dockerfile.celery.template` only for balenaCloud

File: server.py [hunk 3]
```diff
@@ -191,6 +191,42 @@ def cleanup_usb_assets(media_dir='/media'):
                         assets_helper.delete(conn, asset['asset_id'])
 
 
+@celery.task
+def display_status():
+    """Request power status of primary CEC device."""
+    cmd = sh.Command('/usr/bin/cec-client')
```
Comment (rusko124) [inline server.py]: Can be done so `sh.Command('/usr/bin/cec-client')(sh.echo('pow 0'), '-d', '1', '-s')`

File: templates/index.html [hunk 4]
```diff
@@ -304,6 +304,14 @@ <h3 id="modalLabel">Add Asset</h3>
                 <h5 class="d-flex">
                     <b class="justify-content-center align-self-center">Schedule Overview</b>
                     <div class="ml-auto">
+                        <a id="display-on-button" class="btn btn-long btn-outline-dark" href="#">
+                            <i class="fas fa-eye pr-2"></i>
+                            Display on
```
Comment (rusko124) [inline templates/index.html]: Could you add success or failed alert message, please?

Comment (escoand) [reply templates/index.html]: Not sure how to do this. And as https://github.com/Screenly/screenly-ose/pull/1278#issuecomment-564647990 states a successful command does not guarantee the display really switched off. Better would be to call `display_status` afterwards and read the actual status. But we need async UI update for this, but don't know how.

**Comment by escoand** [comment]
@rusko124 I've added scheduling functionality but found no way to update the crontab schedule when running. https://github.com/celery/celery/issues/3493#issuecomment-335997685 states it was added with Celery 4.1.0 and we are using 4.3.0

**Comment by rusko124** [comment]
Good progress @escoand ! Hmm not sure that Celery provides something like this.
In my view how it should work. When user change the scheduling, then system delete the task and add a new one with a new crontab. What do you think?
>I think additionally there needs to be a timezone setting.

Yes, a good point

**Comment by escoand** [comment]
> When user change the scheduling, then system deletes the task and adds a new one with a new crontab.

Apparently, the beat service loads the schedule on start and any change afterwards doesn't do anything.

**Comment by rusko124** [comment]
@escoand look at this https://github.com/sibson/redbeat, I hope it might help. Or It helps you find what you're looking for

**Comment by escoand** [comment]
@rusko124 hmm, I'm doing it exactly as https://stackoverflow.com/a/55603127/3775041 describes, but can't see the periodic sync.

**Comment by rusko124** [comment]
@escoand PR looks good to me! Did you test on a real device?

**Comment by escoand** [comment]
@rusko124 The beat is working but it behaves a bit weird:

```
 screenly-celery  [2020-02-08 09:01:35,871: INFO/Beat] beat: Starting...
 screenly-celery  [2020-02-08 09:01:36,229: DEBUG/Beat] Stored entry: <RedBeatSchedulerEntry: display.on server.display_on() <crontab: 45 6 * * * (m/h/d/dM/MY)>
 screenly-celery  [2020-02-08 09:01:36,246: DEBUG/Beat] Stored entry: <RedBeatSchedulerEntry: cleanup server.cleanup() <freq: 1.00 hour>
 screenly-celery  [2020-02-08 09:01:36,253: DEBUG/Beat] Stored entry: <RedBeatSchedulerEntry: display.off server.display_off() <crontab: 5 9 * * * (m/h/d/dM/MY)>
 screenly-celery  [2020-02-08 09:01:36,259: DEBUG/Beat] Stored entry: <RedBeatSchedulerEntry: cleanup_usb_assets server.cleanup_usb_assets() <freq: 1.00 hour>
 screenly-celery  [2020-02-08 09:01:36,264: DEBUG/Beat] beat: Ticking with max interval->5.00 minutes
 screenly-celery  [2020-02-08 09:01:36,267: DEBUG/Beat] beat: Acquiring lock...
 screenly-celery  [2020-02-08 09:16:36,530: DEBUG/Beat] beat: Extending lock...
 screenly-celery  [2020-02-08 09:16:36,536: DEBUG/Beat] Selecting tasks
 screenly-celery  [2020-02-08 09:16:36,557: INFO/Beat] Loading 1 tasks
 screenly-celery  [2020-02-08 09:16:36,644: INFO/Beat] Scheduler: Sending due task display.off (server.display_off)
 screenly-celery  [2020-02-08 09:16:36,751: DEBUG/Beat] beat: Synchronizing schedule...
 screenly-celery  [2020-02-08 09:16:36,754: DEBUG/Beat] server.display_off sent. id->9ef22021-9594-4538-a8e5-29fa7e202d77
 screenly-celery  [2020-02-08 09:16:36,758: DEBUG/Beat] beat: Waking up in 5.00 minutes.
 screenly-celery  [2020-02-08 09:16:36,759: INFO/MainProcess] Received task: server.display_off[9ef22021-9594-4538-a8e5-29fa7e202d77]  
 screenly-celery  [2020-02-08 09:16:36,764: DEBUG/MainProcess] TaskPool: Apply <function _fast_trace_task at 0x751d5eb0> (args:('server.display_off', '9ef22021-9594-4538-a8e5-29fa7e202d77', {'origin': 'gen25@9ccc31f28dab', 'lang': 'py', 'task': 'server.display_off', 'group': None, 'root_id': '9ef22021-9594-4538-a8e5-29fa7e202d77', u'delivery_info': {u'priority': 0, u'redelivered': None, u'routing_key': 'celery', u'exchange': u''}, 'expires': None, u'correlation_id': '9ef22021-9594-4538-a8e5-29fa7e202d77', 'retries': 0, 'timelimit': [None, None], 'argsrepr': '()', 'eta': None, 'parent_id': None, u'reply_to': '3254029a-f16f-3394-9809-24308524a826', 'shadow': None, 'id': '9ef22021-9594-4538-a8e5-29fa7e202d77', 'kwargsrepr': '{}'}, '[[], {}, {"chord": null, "callbacks": null, "errbacks": null, "chain": null}]', 'application/json', 'utf-8') kwargs:{})
 screenly-celery  [2020-02-08 09:16:36,792: DEBUG/MainProcess] Task accepted: server.display_off[9ef22021-9594-4538-a8e5-29fa7e202d77] pid:24
 screenly-celery  [2020-02-08 09:16:37,751: INFO/ForkPoolWorker-5] Task server.display_off[9ef22021-9594-4538-a8e5-29fa7e202d77] succeeded in 0.970037615s: None
```

The `display.off` was scheduled for 09:05 but gets executed at 09:16.

**Comment by escoand** [comment]
@rusko124 I'm now quite happy with the PR except the described delayed start. It's not happening in every case. Could it be that celery and the beat have problems with using the same redis url?

**Comment by escoand** [comment]
I'd forgotten to mention that the UI has some problems.

- The time picker is not working but don't know why.
- And there is problem when switching from 12h to 24h. The display times need to be reentered *after* the save to get converted to the new format. If not it won't be interpreted correctly.

**Comment by escoand** [comment]
Is there a chance to get it merged finally?

**Pull request closed without merging**